### DISCUSSION TITLE
Injecting uuids during SVG conversion

### DISCUSSION THREAD

**Discussion by bdaniere** [mozman/ezdxf#551] [Q&A]
Hello,
I'm using the DXF to SVG conversion but entities uuid are lost in the process.
One of the problem is sub-geometries. Mathplotlib doesn't support groups (<g>) so we can't associate one uuid to many geometries in the SVG.
Instead of generating a new uuid for each entity, I plan to keep the original uuid (if there is one) and append _1, _2, ... to each sub-geometry so the SVG keeps unique ids.

I can open a PR but your advice may be useful.

**Comment by mbway** [comment]
do you mean like when a lypolyline is broken down into several line segments and curves? in which case I think appending `_1`, `_2` etc is a good idea. I'm not sure what question you're asking, is this approach causing problems?

I think you will have problems with INSERT entities. Especially since you can nest INSERTs inside of INSERTs and the 'virtual entities' inside them have no handle in the dxf. I don't know how best you could identify those. Maybe those entities have handles in the block where they originate but I don't know

**Comment by mozman** [comment]
- All (real) DXF entities stored in a DXF document have a handle (hex-string), accessible by `DXFEntity.dxf.handle`.
- All virtual entities, which are temporarily created entities by copying the original entity or entities created by instantiating the entity class without using a factory method, have no handle until they are bound to a DXF document.
- The property `DXFEntity.uuid` returns an UUID for each DXF entity, even virtual entities. This UUID is created at runtime at the first access of the `uuid` property, it is created by `uuid.uuid4()`. This UUID is not stored in the DXF file and also would not work for virtual entities, because they are not stored in the DXF file.
- The content entities in block references (INSERT) are always virtual entities. They are the result of copying the original BLOCK entities and transforming them to their new position, orientation and size according the parameters of the INSERT entity. These copies do not carry a reference to the source entity in the BLOCK definition.

**Comment by mozman** [comment]
Added a new feature to track the source of copied entities, new properties of `DXFEntity`:

- `source_of_copy` - get immediate source of an entity copy
- `origin_of_copy` - get first non virtual source of an entity copy
- `is_copy` - returns `True` if the entity is a copy

This makes it possible to trace the copying process back to the original entity, even for entities from an INSERT in an INSERT in ...

But you cannot get the source block reference (INSERT) from a virtual entity.

**Comment by mozman** [comment]
Added a feature to trace back the source block references of virtual entities:

New properties of `DXFEntity`:
- `has_source_block_reference` - is `True` if the virtual entity was created by a block reference
- `source_block_reference` - the immediate source block reference (`INSERT`), which created the virtual entity, otherwise ``None``

Application of this property are shown in this tests:
https://github.com/mozman/ezdxf/blob/a37f51cf863af335013609ea20caf1c2f8e827ef/tests/test_04_dxf_high_level_structs/test_414_block_reference_content.py#L416-L471

https://github.com/mozman/ezdxf/blob/a37f51cf863af335013609ea20caf1c2f8e827ef/tests/test_04_dxf_high_level_structs/test_416b_recursive_decompose.py#L118-L191

**Comment by bdaniere** [comment]
For the moment, I have chosen not to retrieve the suffix information (an INSERT containing another INSERT): this information is not useful to me and causes other problems with my methodology.

  ```
def recover_parent_uuid(entity):
      """
      Retrieve the uuid of the entity or parent entity present in d.modelspace()
      Recovery of the uuid could be replaced by the handle
  
      :param entity: (ezdxf.entities.)
      :return: (str) parent's uuid
      """
  
      if not entity.is_virtual and not entity.is_copy:
          # return entity.dxf.handle
          return str(entity.uuid)
  
      if entity.is_copy:
          entity = entity.origin_of_copy
  
      if entity.has_source_block_reference:
          if not entity.source_block_reference.is_virtual:
              entity = entity.source_block_reference
          else:
              while entity.has_source_block_reference:
                  entity = entity.source_block_reference
  
      # return entity.dxf.handle
      return str(entity.uuid)
```

I added the **recover_parent_uuid()** method to _addons.drawing.properties_  to store this information in the properties (in resolve_all()): 

     p.output_id = recover_parent_uuid(entity) 

I would to use this information in the various methods of _addons.drawing.matplotlib_). For example, for _MatplotlibBackend.draw_text()_ :

        patch = PathPatch(
            transformed_path,
            facecolor=properties.color,
            linewidth=0,
            zorder=self._get_z(),
        )
        patch.set_gid(properties.output_id)
        self.ax.add_patch(patch)

        # self.ax.add_patch(
        # PathPatch(
        # transformed_path,
        # facecolor=properties.color,
        # linewidth=0,
        # zorder=self._get_z(),
        # )
        # ) 

I then get the produced svg to merge the paths of the SVG objects according to their id.
I don't know if this approach is the most efficient but it currently works for simple DXF. I give you an example:

[example.zip](https://github.com/mozman/ezdxf/files/7284553/example.zip)
- _drawing_result.svg_ : the result of matplotlib.qsave()
- _merge_path.svg_ : the svg after merge of SVG objects

In addition :
- Not all geometries are taken into account: we still find names such as "patch_4".
- I haven't done any tests on more complex DXFs yet
- the use of suffixes seems to me more complex because we recover the uuids by entities and sub entities ---> I had thought of using hash(str(entity.uuid)) in order to have names of type entity_uuid + "_" + hash(str(sub_entity.uuid)) + "_" hash(str(sub_entity.uuid)). I think there are more relevant solutions
